### ISSUE TITLE
checks for tests?

### ISSUE THREAD

**Issue by maelle** [ropensci-review-tools/pkgcheck#150]
when reading https://github.com/hadley/r-pkgs/pull/777/ I was wondering whether there could be ckecks of

- top-level code in test files
- library calls in test files
- source() calls in test files

**Comment by mpadge** [comment]
Plus important to realise that  many packages only *partly* use `testthat`, and mix those tests with other forms of tests, so actually even that last idea may not be generally feasible?

**Comment by maelle** [comment]
for httptest2 one actually modifies setup.R, not the testthat file?

**Comment by maelle** [comment]
I agree these checks are only for packages using testthat.

For naked code it could be a note. It's useful as it could be code that could be moved to a helper file, and even if not it'd be something interesting for editors/reviewers to notice as it's a tricky thing. Maybe same with library?

**Comment by maelle** [comment]
if a package uses a mix is it possible to isolate which test folder uses testthat?

**Comment by mpadge** [comment]
> for httptest2 one actually modifies setup.R, not the testthat file?

No, the package requires an additional `library` call in `tests/testthat.R`.

> For naked code it could be a note. It's useful as it could be code that could be moved to a helper file, and even if not it'd be something interesting for editors/reviewers to notice as it's a tricky thing. Maybe same with library?

I'm not sure though, because that really would then be complying with the subjective views of effectively one person only (Hadley). There are other very prominent developers and R-core members who do not view that as "best practice". So I'm not even sure we should be systematically/automatically pointing out any divergence from r-pkgs recommendations? That book codifies one approach, but requiring that to be the only approach may be unnecessarily restrictive, I think.

> if a package uses a mix is it possible to isolate which test folder uses testthat?

In principle, yes. Even directly possibly to just use `parse()` on each test file and tease apart sections within single files. But I'd first have to be convinced on the above point.

**Comment by maelle** [comment]
but for testthat it might be fine to try and apply the recs of testthat authors?

If pkgcheck notified about these uses, it'd be interesting as their deviating from the "norm" means they are something relevant to look at while reviewing (compared to boring expectations in the tests).

We could revert if we realize all submissions have a legitimate "different" testthat use. :sweat_smile:

**Comment by maelle** [comment]
I'd still find this useful.

**Comment by mpadge** [comment]
Agreed, and also cross-ref https://lintr.r-lib.org/reference/pkg_testthat_linters.html. Currently incoming #268 will cover some of this now anyway, right? From those ones, we are suggesting:

- [x] [conjunct_test_linter](https://lintr.r-lib.org/reference/conjunct_test_linter.html)
- [ ] [expect_comparison_linter](https://lintr.r-lib.org/reference/expect_comparison_linter.html)
- [ ] [expect_identical_linter](https://lintr.r-lib.org/reference/expect_identical_linter.html)
- [x] [expect_length_linter](https://lintr.r-lib.org/reference/expect_length_linter.html)
- [x] [expect_named_linter](https://lintr.r-lib.org/reference/expect_named_linter.html)
- [x] [expect_not_linter](https://lintr.r-lib.org/reference/expect_not_linter.html)
- [x] [expect_null_linter](https://lintr.r-lib.org/reference/expect_null_linter.html)
- [x] [expect_s3_class_linter](https://lintr.r-lib.org/reference/expect_s3_class_linter.html)
- [x] [expect_s4_class_linter](https://lintr.r-lib.org/reference/expect_s4_class_linter.html)
- [x] [expect_true_false_linter](https://lintr.r-lib.org/reference/expect_true_false_linter.html)
- [x] [expect_type_linter](https://lintr.r-lib.org/reference/expect_type_linter.html)
- [ ] [yoda_test_linter](https://lintr.r-lib.org/reference/yoda_test_linter.html)

Those don't really cover your main initial concerns, though:
- top-level code
- `source()` calls

so they'd still need separate implementations here.
